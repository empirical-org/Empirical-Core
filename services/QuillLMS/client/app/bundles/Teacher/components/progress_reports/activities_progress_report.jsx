import createReactClass from 'create-react-class';
import moment from 'moment';
import queryString from 'query-string';
import React from 'react';

import EmptyStateForReport from './empty_state_for_report.jsx';
import { PROGRESS_REPORTS_SELECTED_CLASSROOM_ID, } from './progress_report_constants';
import ProgressReportFilters from './progress_report_filters.jsx';

import { requestGet, } from '../../../../modules/request/index';
import { ReactTable, } from '../../../Shared/index';
import { getTimeSpent } from '../../helpers/studentReports';
import TableFilterMixin from '../general_components/table/sortable_table/table_filter_mixin';
import LoadingSpinner from '../shared/loading_indicator.jsx';


export default createReactClass({
  displayName: 'activities_progress_report',

  mixins: [TableFilterMixin],

  getInitialState: function() {
    return {
      loadingFilterOptions: true,
      currentPage: 0,
      csvData: [],
      currentSort: {sort_param: 'completed_at', sort_descending: true},
      premium: false
    }
  },

  componentDidMount: function() {
    this.fetchData();
  },

  requestParams: function() {
    const { currentPage, currentFilters, currentSort, filtersLoaded, } = this.state
    let params = Object.assign({},
      { page: currentPage + 1 },
      currentFilters,
      currentSort
    );

    if(filtersLoaded) {
      params = Object.assign(params, {without_filters: true});
    }

    return params;
  },

  fetchData: function() {
    const { filtersLoaded, } = this.state
    this.setState({ loadingNewTableData: true });
    requestGet(
      queryString.stringifyUrl({ url: `${process.env.DEFAULT_URL}/teachers/progress_reports/activity_sessions.json`, query: this.requestParams()}),
      (data) => {
        const classroomFilters = this.getFilterOptions(data.classrooms, 'name', 'id', 'All classes')
        const studentFilters = this.getFilterOptions(data.students, 'name', 'id', 'All students')
        const unitFilters = this.getFilterOptions(data.units, 'name', 'id', 'All activity packs')

        const selectedClassroomId = window.localStorage.getItem(PROGRESS_REPORTS_SELECTED_CLASSROOM_ID)

        if (!filtersLoaded && selectedClassroomId && classroomFilters.find(c => Number(c.value) === Number(selectedClassroomId))) {
          const newState = {
            classroomFilters,
            studentFilters,
            unitFilters,
            filtersLoaded: true,
            currentFilters: { classroom_id: Number(selectedClassroomId), }
          }
          this.setState(newState, () => this.fetchData())
        } else {
          let newState = {
            loadingFilterOptions: false,
            loadingNewTableData: false,
            results: data.activity_sessions,
            numPages: data.page_count,
            classroomFilters,
            studentFilters,
            unitFilters,
          };

          if (!filtersLoaded) {
            newState = Object.assign(newState, {
              filtersLoaded: true
            })
          }

          this.setState(newState, () => {
            const { classroomFilters, selectedStudent, studentFilters, selectedUnit, unitFilters, } = this.state
            const selectedClassroom = classroomFilters.find(c => Number(c.value) === Number(selectedClassroomId)) || classroomFilters[0]
            this.setState({
              selectedClassroom,
              selectedStudent: selectedStudent || studentFilters[0],
              selectedUnit: selectedUnit || unitFilters[0]
            })
          });
        }
      }
    );
  },

  canViewReport: function() {
    const { premiumStatus, } = this.props
    return premiumStatus === 'paid' || premiumStatus === 'trial'
  },

  columnDefinitions: function() {
    const { studentFilters, } = this.state
    const cellClassName = this.nonPremiumBlur()
    // Student, Date, Activity, Score, Standard, Tool
    return [
      {
        Header: 'Student',
        accessor: 'student_id',
        resizeable: false,
        Cell: ({ row }) => <span className={cellClassName}>{studentFilters.find(student => student.value == row.original.student_id)?.name}</span>,
        maxWidth: 200
      },
      {
        Header: 'Date',
        accessor: 'completed_at',
        resizeable: false,
        Cell: ({row}) => moment.unix(Number(row.original.completed_at)).format('M/D/YY'),
        maxWidth: 90
      },
      {
        Header: 'Activity',
        accessor: 'activity_name',
        resizeable: false,
        Cell: ({ row }) => <span className="show-overflow">{row.original.activity_name}</span>
      },
      {
        Header: 'Score',
        accessor: 'percentage',
        resizeable: false,
        Cell: ({row}) => {
          const score = row.original.percentage >= 0 ? `${Math.round(row.original.percentage * 100)}%` : 'Completed'
          return <span className={cellClassName}>{score}</span>
        },
        maxWidth: 90
      },
      {
        Header: 'Time spent',
        accessor: 'timespent',
        resizeable: false,
        Cell: ({ row }) => <span className={cellClassName}>{getTimeSpent(row.original.timespent)}</span>,
        maxWidth: 90
      },
      {
        Header: 'Standard',
        accessor: 'standard',
        resizeable: false,
        Cell: ({row}) => row.original.standard ? row.original.standard.split(' ')[0] : '',
        maxWidth: 90
      },
      {
        Header: 'Tool',
        accessor: 'activity_classification_name',
        resizeable: false,
        maxWidth: 150
      }
    ];
  },

  selectClassroom: function(classroom) {
    this.setState({ selectedClassroom: classroom, });
    window.localStorage.setItem(PROGRESS_REPORTS_SELECTED_CLASSROOM_ID, classroom.value)
    this.filterByField('classroom_id', classroom.value, this.onFilterChange);
  },

  selectStudent: function(student) {
    this.setState({ selectedStudent: student, });
    this.filterByField('student_id', student.value, this.onFilterChange);
  },

  selectUnit: function(unit) {
    this.setState({ selectedUnit: unit, });
    this.filterByField('unit_id', unit.value, this.onFilterChange);
  },

  onFilterChange: function() {
    this.setState({loading: true, currentPage: 0}, this.fetchData);
  },

  reactTableSortedChange: function(state, instance) {
    if (!state.length) { return }
    const { currentSort, } = this.state
    const newSort = {
      sort_param: state[0].id,
      sort_descending: state[0].desc,
    }

    if (_.isEqual(currentSort, newSort)) { return }

    this.setState({
      // We want to revert to the first page if changing anything other than the page.
      currentPage: 0,
      currentSort: {
        sort_param: state[0].id,
        sort_descending: state[0].desc,
      }
    }, this.fetchData);
  },

  reactTablePageChange: function(newPage) {
    const { currentPage, } = this.state
    if (currentPage === newPage) { return }
    this.setState({
      currentPage: newPage,
    }, this.fetchData);
  },

  tableOrEmptyMessage: function(){
    const {
      results,
      loadingNewTableData,
      currentPage,
      numPages,
      classroomFilters,
      selectedClassroom,
      selectedStudent,
      selectedUnit,
      studentFilters,
      unitFilters,
      currentSort,
    } = this.state
    let tableOrEmptyMessage
    if (results.length) {
      tableOrEmptyMessage = (<ReactTable
        className='progress-report'
        columns={this.columnDefinitions()}
        currentPage={currentPage}
        data={results}
        defaultPageSize={Math.min(results.length, 25)}
        defaultSorted={[{id: currentSort.sort_param, desc: currentSort.sort_descending}]}
        loading={loadingNewTableData}
        manualPageCount={numPages}
        manualPagination={true}
        manualSortBy={true}
        onPageChange={this.reactTablePageChange}
        onSortedChange={this.reactTableSortedChange}
        resizable={false}
        showPageSizeOptions={false}
        showPagination={true}
        showPaginationBottom={true}
        showPaginationTop={false}
      />)
    } else {
      tableOrEmptyMessage = <EmptyStateForReport />
    }
    return (
      <div>
        <ProgressReportFilters
          classroomFilters={classroomFilters}
          filterTypes={['unit', 'classroom', 'student']}
          selectClassroom={this.selectClassroom}
          selectedClassroom={selectedClassroom}
          selectedStudent={selectedStudent}
          selectedUnit={selectedUnit}
          selectStudent={this.selectStudent}
          selectUnit={this.selectUnit}
          studentFilters={studentFilters}
          unitFilters={unitFilters}
        />
        {tableOrEmptyMessage}
      </div>
    )
  },

  renderFiltersAndTable: function() {
    const { loadingFilterOptions, loadingNewTableData, } = this.state
    if(loadingFilterOptions || loadingNewTableData) {
      return <LoadingSpinner />
    }
    return (this.tableOrEmptyMessage())
  },

  paramsToQueryString: function(){
    // converts the params object to a query string
    // https://stackoverflow.com/a/35416293/2812720
    const obj = this.requestParams()
    return `?${Object.keys(obj).map(k => `${encodeURIComponent(k)}=${encodeURIComponent(obj[k])}`).join('&')}`;
  },

  nonPremiumBlur: function() {
    return this.canViewReport() ? '' : 'non-premium-blur';
  },

  downloadReport: function() {
    if(this.canViewReport()) {
      return window.open(`/teachers/progress_reports/activity_sessions.csv${this.paramsToQueryString()}`);
    }
    alert('Downloadable reports are a Premium feature. You can visit Quill.org/premium to upgrade now!');
  },

  render: function() {
    return (
      <div className='progress-reports-2018 data-export'>
        <div className='meta-overview flex-row space-between'>
          <div className='header-and-info'>
            <h1>Data Export</h1>
            <p>You can export the data as a CSV file by filtering for the classrooms, activity packs, or students you would like to export and then pressing "Download Report."</p>
          </div>
          <div className='csv-and-how-we-grade'>
            <button className='quill-button medium primary contained focus-on-light' onClick={this.downloadReport} style={{display: 'block'}}>Download Report</button>
            <a className='how-we-grade' href="https://support.quill.org/activities-implementation/how-does-grading-work">How We Grade<i className="fas fa-long-arrow-alt-right" /></a>
          </div>
        </div>
        {this.renderFiltersAndTable()}
      </div>
    );
  },
});
