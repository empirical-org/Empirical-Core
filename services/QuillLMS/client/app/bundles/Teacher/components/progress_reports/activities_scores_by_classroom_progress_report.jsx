import React from 'react'
import request from 'request'
import _ from 'underscore'
import queryString from 'query-string';
import ReactTooltip from 'react-tooltip';
import ReactTable from 'react-table'
import 'react-table/react-table.css'
import moment from 'moment'

import EmptyStateForReport from './empty_state_for_report'
import CSVDownloadForProgressReport from './csv_download_for_progress_report.jsx'

import ItemDropdown from '../general_components/dropdown_selectors/item_dropdown'
import LoadingSpinner from '../shared/loading_indicator.jsx'
import {sortByLastName, sortFromSQLTimeStamp} from '../../../../modules/sortingMethods.js'
import { getTimeSpent } from '../../helpers/studentReports';

const showAllClassroomKey = 'All classes'

export class ActivitiesScoresByClassroomProgressReport extends React.Component {
  constructor() {
    super()
    this.state = {
      loading: true,
      errors: false,
      selectedClassroom: showAllClassroomKey
    }
  }

  componentDidMount() {
    request.get({
      url: `${process.env.DEFAULT_URL}/api/v1/progress_reports/activities_scores_by_classroom_data`
    }, (e, r, body) => {
      const data = JSON.parse(body).data
      const classroomsData = data;
      // gets unique classroom names
      const classroomNames = Array.from(new Set(classroomsData.map(row => row.classroom_name)))
      classroomNames.unshift(showAllClassroomKey)

      const newState = {loading: false, errors: body.errors, classroomsData, classroomNames}
      const selectedClassroomId = queryString.parse(window.location.search).classroom_id

      if (selectedClassroomId) {
        const selectedClassroom = classroomsData.find(c => Number(c.classroom_id) === Number(selectedClassroomId))
        newState.selectedClassroom = selectedClassroom.classroom_name || showAllClassroomKey
      }
      this.setState(newState);
    });
  }

  columns() {
    return ([
      {
        Header: 'Student',
        accessor: 'name',
        resizable: false,
        sortMethod: sortByLastName,
        Cell: row => (<a className='row-link-disguise underlined' href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`}>
          {row.original.name}
        </a>),
      }, {
        Header: "Activities completed",
        accessor: 'activity_count',
        resizable: false,
        minWidth: 110,
        Cell: row => (<a className='row-link-disguise' href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`}>
          {Number(row.original.activity_count)}
        </a>),
      }, {
        Header: "Overall score",
        accessor: 'average_score',
        resizable: false,
        minWidth: 80,
        Cell: row => {
          const value = Math.round(parseFloat(row.original.average_score) * 100);
          return (<a className='row-link-disguise' href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`}>
            {isNaN(value) ? '--' : value + '%'}
          </a>)
        }
      },
      {
        Header: "Time spent",
        accessor: 'timespent',
        resizable: false,
        minWidth: 80,
        Cell: row => {
          const value = row.original.timespent;
          return (<a className='row-link-disguise' href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`}>
            {isNaN(value) ? '--' : getTimeSpent(value)}
          </a>)
        }
      },
      {
				Header: "Last active",
				accessor: 'last_active',
				resizable: false,
        minWidth: 90,
        Cell: row => (<a className='row-link-disguise' href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`}>
          {row.original.last_active ? moment(row.original.last_active).format("MM/DD/YYYY") : <span />}
        </a>),
				sortMethod: sortFromSQLTimeStamp,
			},
			{
        Header: "Class",
        accessor: 'classroom_name',
        resizable: false,
        Cell: row => (
          <div className="activate-tooltip" data-for="classroom-tooltip" data-tip={`<p>${row.original.classroom_name}</p>`}>
            <ReactTooltip className="react-tooltip-custom" effect="solid" html id="classroom-tooltip" multiline type="light" />
            <a className="tooltip-trigger row-link-disguise" href={`/teachers/progress_reports/student_overview?classroom_id=${row.original.classroom_id}&student_id=${row.original.student_id}`} rel='noreferrer noopener' target="_blank">{row.original.classroom_name}</a>
          </div>
        )
      }
    ])
  }

  filteredClassroomsData() {
    const { classroomsData, selectedClassroom } = this.state;
    if (selectedClassroom === showAllClassroomKey) {
      return classroomsData
    }
    return classroomsData.filter((row) => row.classroom_name === selectedClassroom)
  }

  formatDataForCSV(data) {
    const csvData = [
      ['Classroom Name', 'Student Name', 'Average Score', 'Activity Count']
    ]
    data.forEach((row) => {
      csvData.push([
        row['classroom_name'], row['name'], (row['average_score'] * 100).toString() + '%',
        row['activity_count']
      ])
    })
    return csvData
  }

  switchClassrooms = classroom => {
    const { classroomsData, } = this.state
    const classroomRecord = classroomsData.find(c => c.classroom_name === classroom)
    if (classroomRecord) {
      window.history.pushState({}, '', `${window.location.pathname}?classroom_id=${classroomRecord.classroom_id}`);
    } else {
      window.history.pushState({}, '', window.location.pathname);
    }
    this.setState({selectedClassroom: classroom})
  };

  tableOrEmptyMessage(filteredClassroomsData){
    if (filteredClassroomsData.length) {
      return (<div key={`${filteredClassroomsData.length}-length-for-activities-scores-by-classroom`}>
        <ReactTable
          className='progress-report'
          columns={this.columns()}
          data={filteredClassroomsData}
          defaultPageSize={filteredClassroomsData.length}
          defaultSorted={[{id: 'last_active', desc: true}]}
          showPageSizeOptions={false}
          showPagination={false}
          showPaginationBottom={false}
          showPaginationTop={false}
        />
      </div>)
    } else {
      return <EmptyStateForReport />
    }
  }

  render() {
    const { errors, loading, selectedClassroom, classroomNames } = this.state;
    if (errors) {
      return <div className='errors'>{errors}</div>
    }
    if (loading) {
      return <LoadingSpinner />
    }
    const filteredClassroomsData = this.filteredClassroomsData()
    return (
      <div className='activities-scores-by-classroom progress-reports-2018'>
        <div className="meta-overview flex-row space-between">
          <div className='header-and-info'>
            <h1>Activity Scores</h1>
            <p>View the overall average score for each student in an active classroom. Click on a studentâ€™s name to see a report of each individual activity and print it as a PDF. You can print this report by downloading a PDF file or export this data by downloading a CSV file.</p>
          </div>
          <div className='csv-and-how-we-grade'>
            <CSVDownloadForProgressReport data={this.formatDataForCSV(filteredClassroomsData)} key={`${selectedClassroom} report button`} />
            <a className='how-we-grade' href="https://support.quill.org/activities-implementation/how-does-grading-work">How We Grade<i className="fas fa-long-arrow-alt-right" /></a>
          </div>
        </div>
        <div className='dropdown-container'>
          <ItemDropdown callback={this.switchClassrooms} items={classroomNames} selectedItem={selectedClassroom} />
        </div>
        {this.tableOrEmptyMessage(filteredClassroomsData)}
      </div>
    )
  }
}

export default ActivitiesScoresByClassroomProgressReport
