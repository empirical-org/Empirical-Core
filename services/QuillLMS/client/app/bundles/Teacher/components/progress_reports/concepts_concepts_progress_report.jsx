// The progress report shows all concepts for a given student.
import React from 'react'
import CSVDownloadForProgressReport from './csv_download_for_progress_report.jsx'

import { requestGet, } from '../../../../modules/request/index'
import { ReactTable, } from '../../../Shared/index'
import userIsPremium from '../modules/user_is_premium'
import LoadingSpinner from '../shared/loading_indicator.jsx'

export default class IndividualStudentConceptReport extends React.Component {

  constructor(props) {
    super()
    this.state = {
      loading: true,
      errors: false,
      userIsPremium: userIsPremium()
    }
  }

  componentDidMount() {
    requestGet(
      `${import.meta.env.VITE_DEFAULT_URL}/${this.props.sourceUrl}`,
      (body) => {
        this.setState({
          loading: false,
          errors: body.errors,
          reportData: body.concepts,
          studentName: body.student.name
        });
      }
    )
  }

  columns() {
    const blurIfNotPremium = this.state.userIsPremium ? null : 'non-premium-blur'
    return ([
      {
        Header: 'Category',
        accessor: 'level_2_concept_name',
        resizable: false,
        maxWidth: 200
      }, {
        Header: 'Name',
        accessor: 'concept_name',
        className: 'show-overflow',
        resizable: false,
      }, {
        Header: 'Questions',
        accessor: 'total_result_count',
        resizable: false,
        maxWidth: 120
      }, {
        Header: 'Correct',
        accessor: 'correct_result_count',
        resizable: false,
        Cell: ({ row }) => <span className={blurIfNotPremium}>{row.original.correct_result_count}</span>,
        maxWidth: 105
      }, {
        Header: 'Incorrect',
        accessor: 'incorrect_result_count',
        resizable: false,
        Cell: ({ row }) => <span className={blurIfNotPremium}>{row.original.incorrect_result_count}</span>,
        maxWidth: 115
      }, {
        Header: 'Percentage',
        accessor: 'percentage',
        resizable: false,
        Cell: ({ row }) => <span className={blurIfNotPremium}>{row.original.percentage + '%'}</span>,
        maxWidth: 120
      },
    ])
  }

  render() {
    if (this.state.loading || !this.state.reportData) {
      return <LoadingSpinner />
    }
    return (
      <div className='progress-reports-2018 concept-student-concepts' concept-student-concepts>
        <div className="meta-overview flex-row space-between">
          <div className='header-and-info flex-row vertically-centered'>
            <h1><span>Concept Results:</span> {this.state.studentName}</h1>
            <p>You can print this report by downloading a PDF file or export this data by downloading a CSV file.</p>
          </div>
          <div className='csv-and-how-we-grade'>
            <CSVDownloadForProgressReport data={this.state.reportData} studentName={this.state.studentName} />
            <a className='how-we-grade' href="https://support.quill.org/activities-implementation/how-does-grading-work">How We Grade<i className="fas fa-long-arrow-alt-right" /></a>
          </div>
        </div>
        <div className='student-name'>{this.state.studentName}</div>
        <div key={`concepts-concepts-progress-report-length-${this.state.reportData.length}`}>
          <ReactTable
            className='progress-report has-green-arrow'
            columns={this.columns()}
            data={this.state.reportData}
            defaultSorted={[{
              id: 'total_result_count',
              desc: true
            }
            ]}
          />
        </div>
      </div>
    )
  }

};
