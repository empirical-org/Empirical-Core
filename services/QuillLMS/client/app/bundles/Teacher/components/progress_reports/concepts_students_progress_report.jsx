// The progress report showing all students in a given classroom
// along with their result counts.
import React from 'react'
import queryString from 'query-string';

import CSVDownloadForProgressReport from './csv_download_for_progress_report.jsx'
import { PROGRESS_REPORTS_SELECTED_CLASSROOM_ID, } from './progress_report_constants'
import EmptyStateForReport from './empty_state_for_report'

import {sortTableByLastName} from '../../../../modules/sortingMethods.js'
import LoadingSpinner from '../shared/loading_indicator.jsx'
import ItemDropdown from '../general_components/dropdown_selectors/item_dropdown'
import userIsPremium from '../modules/user_is_premium'
import { ReactTable, } from '../../../Shared/index'
import { requestGet, } from '../../../../modules/request/index'

const showAllClassroomKey = 'All Classrooms'

export default class ConceptsStudentsProgressReport extends React.Component {
  constructor(props) {
    super()
    this.state = {
      loading: true,
      errors: false,
      selectedClassroomId: showAllClassroomKey,
      userIsPremium: userIsPremium()
    }
  }

  componentDidMount() {
    const that = this;
    requestGet(
      `${import.meta.env.VITE_DEFAULT_URL}/${this.props.sourceUrl}`,
      (body) => {
        const parsedClassrooms = this.parseClassrooms(body.classrooms_with_student_ids)
        const dropdownClassrooms = parsedClassrooms.dropdownClassrooms;
        const classroomsWithStudentIds = parsedClassrooms.classroomsWithStudentIds

        const newState = {loading: false, errors: body.errors, reportData: body.students, filteredReportData: body.students, dropdownClassrooms, classroomsWithStudentIds}

        const selectedClassroomId = queryString.parse(window.location.search).classroom_id
        const localStorageSelectedClassroomId = window.localStorage.getItem(PROGRESS_REPORTS_SELECTED_CLASSROOM_ID)

        if (selectedClassroomId) {
          newState.selectedClassroomId = selectedClassroomId
        } else if (localStorageSelectedClassroomId && dropdownClassrooms.find(c => Number(c.id) === Number(localStorageSelectedClassroomId))) {
          newState.selectedClassroomId = Number(localStorageSelectedClassroomId)
        }
        that.setState(newState, this.filterReportData);
      }
    )
  }

  columns() {
    const { userIsPremium, } = this.state
    const blurIfNotPremium = userIsPremium ? null : 'non-premium-blur'
    const cellClassName = `row-link-disguise ${blurIfNotPremium}`
    return ([
      {
        Header: 'Student',
        accessor: 'name',
        resizable: false,
        sortType: sortTableByLastName,
        width: 174,
        Cell: ({row}) => (
          <a href={row.original['concepts_href']}>{row.original['name']}</a>
        )
      }, {
        Header: 'Questions',
        accessor: 'total_result_count',
        resizable: false,
        Cell: ({row}) => (
          <a className="row-link-disguise" href={row.original['concepts_href']}>{row.original['total_result_count']}</a>
        )
      }, {
        Header: 'Correct',
        accessor: 'correct_result_count',
        resizable: false,
        Cell: ({row}) => (
          <a className={cellClassName} href={row.original['concepts_href']}>{row.original['correct_result_count']}</a>
        )
      }, {
        Header: 'Incorrect',
        accessor: 'incorrect_result_count',
        resizable: false,
        Cell: ({row}) => (
          <a className={cellClassName} href={row.original['concepts_href']}>{row.original['incorrect_result_count']}</a>
        )
      }, {
        Header: 'Percentage',
        accessor: 'percentage',
        resizable: false,
        Cell: ({row}) => (
          <a className={cellClassName} href={row.original['concepts_href']}>{row.original['percentage']}%</a>
        )
      }, {
        Header: "",
        accessor: 'green_arrow',
        resizable: false,
        sortable: false,
        maxWidth: 80,
        Cell: ({row}) => (
          <a className='green-arrow' href={row.original['concepts_href']}>
            <img alt="" src="https://assets.quill.org/images/icons/chevron-dark-green.svg" />
          </a>
        )
      }
    ])
  }

  filterReportData(){
    const { selectedClassroomId, reportData, classroomsWithStudentIds, } = this.state
    let filteredReportData;
    if (selectedClassroomId === showAllClassroomKey) {
      // because we are showing all classrooms, we show all data
      filteredReportData = reportData;
    } else {
      const validStudentIds = classroomsWithStudentIds[selectedClassroomId]
      filteredReportData = reportData.filter((student)=> validStudentIds.includes(student.id))
    }
    this.setState({ filteredReportData, updatingReportData: false})
  }

  keysToOmit(){
    return ['concepts_href']
  }

  parseClassrooms(classrooms){
    const classroomsWithStudentIds = {}
    const dropdownClassrooms = [{id: showAllClassroomKey, name: showAllClassroomKey}];
    classrooms.forEach((c)=>{
      classroomsWithStudentIds[c.id] = c.student_ids;
      dropdownClassrooms.push({id: c.id, name: c.name})
    })
    return {dropdownClassrooms, classroomsWithStudentIds }
  }

  switchClassrooms = classroom => {
    const { dropdownClassrooms, } = this.state
    const classroomRecord = dropdownClassrooms.find(c => c.id === classroom.id)
    window.localStorage.setItem(PROGRESS_REPORTS_SELECTED_CLASSROOM_ID, classroom.id)

    if (classroomRecord && classroomRecord.id !== showAllClassroomKey) {
      window.history.pushState({}, '', `${window.location.pathname}?classroom_id=${classroomRecord.id}`);
    } else {
      window.history.pushState({}, '', window.location.pathname);
    }
    this.setState({selectedClassroomId: classroom.id, updatingReportData: true, }, this.filterReportData)
  };

  tableOrEmptyMessage(){
    const { filteredReportData, } = this.state
    if (filteredReportData.length) {
      return (
        <div key={`concept-progress-report-length-${filteredReportData.length}`}>
          <ReactTable
            className='progress-report has-green-arrow'
            columns={this.columns()}
            data={filteredReportData}
            defaultSorted={[{
              id: 'total_result_count',
              desc: true
            }
            ]}
          />
        </div>
      )
    } else {
      return <EmptyStateForReport />
    }
  }

  render() {
    const { loading, reportData, dropdownClassrooms, selectedClassroomId, filteredReportData, updatingReportData, } = this.state
    if (loading || !reportData) {
      return <LoadingSpinner />
    }
    const changeValues = [{key: 'percentage', function: ((num)=>num.toString() + '%')}]

    const selectedClassroom = dropdownClassrooms.find(c => String(c.id) === String(selectedClassroomId))

    return (
      <div className='progress-reports-2018'>
        <div className="meta-overview flex-row space-between">
          <div className='header-and-info'>
            <h1>Concept Results</h1>
            <p>Each question on Quill targets a specific writing concept. This report shows the number of times the student correctly or incorrectly used the targeted concept to answer the question. You can see a student’s results on each concept by clicking on the student’s name. You can print this report by downloading a PDF file or export this data by downloading a CSV file.</p>
          </div>
          <div className='csv-and-how-we-grade'>
            <CSVDownloadForProgressReport data={filteredReportData} key={`reports are ready ${updatingReportData}`} keysToOmit={this.keysToOmit()} valuesToChange={changeValues} />
            <a className='how-we-grade' href="https://support.quill.org/activities-implementation/how-does-grading-work">How We Grade<i className="fas fa-long-arrow-alt-right" /></a>
          </div>
          <div className='dropdown-container'>
            <ItemDropdown callback={this.switchClassrooms} items={dropdownClassrooms} selectedItem={selectedClassroom} />
          </div>
        </div>
        {this.tableOrEmptyMessage()}
      </div>
    )
  }
};
