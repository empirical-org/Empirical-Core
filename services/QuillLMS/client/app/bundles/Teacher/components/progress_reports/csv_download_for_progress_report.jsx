import React from 'react'
import {CSVDownload, CSVLink} from 'react-csv'
import userIsPremium from '../modules/user_is_premium'
import _ from 'underscore'
import _l from 'lodash'

function formatData(data) {
  return data.map(row => {
    delete row['Concept Id']
    return row
  })
}

const unclickedDownloadBtnClass = "btn button-green"
const clickedDownloadBtnClass = "quill-button medium primary contained focus-on-light"
const printBtnClass = "print-button"
const printImgClass = "print-img"

export default class CSVDownloadForProgressReports extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      userIsPremium: userIsPremium(),
      showDropdown: false
    }

    document.addEventListener('click', this.closeDropdownIfOpen.bind(this))
  }


  componentDidMount() {
    this.cleanData(this.props.data)
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.closeDropdownIfOpen.bind(this));
  }

  getRidOfCamelCase(row){
    const keys = Object.keys(row)
    keys.forEach(oldKey=>{
      const newKey = _l.startCase(oldKey)
      row[newKey] = row[oldKey];
      delete row[oldKey];
    })
  }

  changeValues(row) {
    this.props.valuesToChange.forEach(keyFunction => {
      row[keyFunction.key] = keyFunction.function(row[keyFunction.key])
    })
  }

  formatFilename(studentName) {
    return `concept_results_for_${studentName}`
  }

  cleanData(data) {
    let finalValue
    if (!Array.isArray(data[0])) {
      const copiedData = JSON.parse(JSON.stringify(data))
      let newData = []
      copiedData.forEach(row => {
        if (this.props.keysToOmit) {
          row = this.omitKeys(row)
        }
        if (this.props.valuesToChange) {
          this.changeValues(row)
        }
        if (!this.props.preserveCasing) {
          this.getRidOfCamelCase(row)
        }
        newData.push(row)
      })
      finalValue = newData
    } else {
      finalValue = data
    }
    this.setState({data: finalValue})
  }

  closeDropdownIfOpen(e) {
    const { className, } = this.props
    const ignoreClasses = [printBtnClass, printImgClass, clickedDownloadBtnClass, unclickedDownloadBtnClass, className]
    if (this.state.showDropdown && !ignoreClasses.includes(e.target.classList.value)) {
      this.setState({ showDropdown: false})
    }
  }

  handleClick() {
    alert('Downloadable reports are a Premium feature. Visit Quill.org/premium to upgrade now!')
  }

  omitKeys(row) {
    return _.omit(row, this.props.keysToOmit)
  }

  toggleDropdown = () => {
    this.setState({ showDropdown: !this.state.showDropdown })
  };

  render() {
    if (this.state.userIsPremium && this.props.data) {
      let dropdown,
        style,
        c
      if (this.state.showDropdown) {
        style = {
          // border: "solid 1px #00c2a2",
          boxShadow: "0 0 0 1px #00c2a2",
          backgroundColor: '#fff',
          color: '#00c2a2'
        }
        dropdown = (
          <div className='dropdown-content' onBlur={this.toggleDropdown}>
            Download report as:
            <span className='pull-right'>
              <div>
                <i aria-hidden="true" className="fa  fa-caret-up gray-caret" />
              </div>
              <div>
                <i aria-hidden="true" className="fas fa-caret-up " />
              </div>
            </span>

            <div className='button-wrapper'>

              <button className={printBtnClass} onClick={window.print}>
                <img
                  alt="print"
                  className={printImgClass}
                  src="https://assets.quill.org/images/icons/download-report-premium.svg"
                />PDF
              </button>

              <CSVLink
                data={formatData(this.state.data)}
                filename={this.formatFilename(this.props.studentName)}
                target="_blank"
              >
                <button>
                  <img alt="csv" src="https://assets.quill.org/images/icons/download-report-premium-csv.svg" />
                  CSV
                </button>
              </CSVLink>

            </div>
          </div>
        )
      }
      return (
        <div className='download-button-wrapper'>
          <button className={clickedDownloadBtnClass} onClick={this.toggleDropdown} style={style}>{this.props.buttonCopy || "Download Report"}</button>
          {dropdown}
        </div>
      )
    } else {
      return (
        <button
          className={this.props.className || unclickedDownloadBtnClass}
          onClick={this.handleClick}
          style={{display: 'block'}}
        >
          {this.props.buttonCopy || "Download Report"}
        </button>
      )
    }
  }
}
