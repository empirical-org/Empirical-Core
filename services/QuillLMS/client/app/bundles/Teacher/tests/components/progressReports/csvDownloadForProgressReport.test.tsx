import { mount } from 'enzyme';
import * as React from 'react';

import { CSVDownloadForProgressReport } from '../../../components/progress_reports/csv_download_for_progress_report';

jest.mock('../../../components/modules/user_is_premium', () => jest.fn(() => true));

describe('ActivitySurvey Component', () => {
  const mockData = [
    ['Classroom Name', 'Student Name', 'School Name', 'Teacher Name', 'Average Score', 'Activity Count', 'Last Active'],
    ['Test Class', 'Jen Penn', 'Drag Rage Academy', 'Jane Fonda', '99%', 2, '2022-03-29 15:13:19']
  ];
  let component = mount(<CSVDownloadForProgressReport data={mockData} />);

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
