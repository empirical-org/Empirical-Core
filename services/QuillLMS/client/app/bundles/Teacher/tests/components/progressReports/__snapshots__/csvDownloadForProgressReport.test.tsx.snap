// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivitySurvey Component should match snapshot 1`] = `
<CSVDownloadForProgressReport
  data={
    Array [
      Array [
        "Classroom Name",
        "Student Name",
        "School Name",
        "Teacher Name",
        "Average Score",
        "Activity Count",
        "Last Active",
      ],
      Array [
        "Test Class",
        "Jen Penn",
        "Drag Rage Academy",
        "Jane Fonda",
        "99%",
        2,
        "2022-03-29 15:13:19",
      ],
    ]
  }
>
  <div
    className="download-button-wrapper"
  >
    <button
      className="quill-button medium primary contained focus-on-light"
      onClick={[Function]}
    >
      Download Report
    </button>
  </div>
</CSVDownloadForProgressReport>
`;
